### PR TITLE
Add JSON support to UUID

### PR DESCRIPTION
**Pull request by lachlan**
Adds the following methods to support for `UUID`s being used directly in JSON mappings, as per issue #5548:

* `UUID.new(JSON::PullParser)`
* `UUID#to_json(JSON::Builder)`

### CODE REVIEW

**Comment by bew** [comment]
Will there be YAML support too?

(would be great to have generic text-based serialization to avoid repeating & 'bloating' the std with code for _all_ protocols)

**Comment by asterite** [comment]
Well, one can easily write a converter for such stuff. Here are wrote two: https://github.com/asterite/wanikani/blob/master/src/wanikani/models.cr#L125-L152

I personally don't think we need to bloat the std with these methods. Writing a converter is super simple.

**Comment by asterite** [comment]
And, to be honest, when you are consuming ids from an API, you are interested in their string format, not in their bytes as represented by the `UUID` struct. In fact, I'm not sure why we have such struct in the std... I know it's a bit more efficient than having just a string for it, but what's the point? You can't do anything with it unless you convert it to a string (I think). Or what's their use case? In Ruby if I do `SecureRandom.uuid` I get a string, not some funky object.

**Comment by asterite** [comment]
And in fact, doing `UUID.random.to_s` allocates twice (first for the bytes, then for the string), while Ruby's `SecureRandom.uuid` allocates just once (for the string).

**Comment by lachlan** [comment]
@asterite the main benefit for me is that a `UUID` is more strongly-typed than a `String`; it has a specific format and properties that a `String` does not have. Not all `String`s are valid `UUID`s.

@bew yes YAML support would be good, as would XML, and whatever other future serialization formats come along. I also agree and don't think the current approach used here is scalable. I'd also prefer a generic text serialization interface, and changing the JSON and YAML implementations to support that interface as a fallback, so that any class that implements the interface gets JSON and YAML (and XML etc) support for free.

**Comment by straight-shoota** [comment]
@asterite Where would you call `UUID.random.to_s` to get a string object? In most cases it'll be written to an IO anyway and then `UUID.random.to_s(io)` is the idiomatic way which doesn't allocate and UUID is overall because it's a struct.

**Comment by asterite** [comment]
To generate an id for an object and store it in a database or a model. I do that **all the time** in my workplace. I never need to see an uuid as an UUID object, always as a string.

**Comment by asterite** [comment]
Now I wonder about use cases where you don't need an uuid as a string. What are they?

**Comment by RX14** [comment]
@asterite plus, it doesn't really allocate because UUID is a struct. It's probably faster than a string in typical cases.

UUID provides more flexibility for the (admittedly rare) case that you need to introspect or modify a UUID. Not all UUIDs are filled with random bytes, although that's typically how they're used nowadays.

**Comment by asterite** [comment]
Ooooh... right! I didn't see it was implemented as a wrapper of `StaticArray(UInt8, 16)`. I thought it allocated those bytes on the heap. You can disregard my previous comments then.

About YAML and other formats: if you can find a way to have something be magically serialized to every possible serialization format, I'm all for it.

**Comment by asterite** [comment]
I rebased the code via github but it still gives conflicts plus it's missing running the formatter.

File: src/uuid/json.cr [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+# require "uuid/json"
+#
+# class Example
+#   JSON.mapping id: UUID
+# end
+#
+# example = Example.from_json(%({"id": "ba714f86-cac6-42c7-8956-bcf5105e1b81"}))
+#
+# uuid = UUID.new("87b3042b-9b9a-41b7-8b15-a93d3f17025e")
+# uuid.to_json # => "87b3042b-9b9a-41b7-8b15-a93d3f17025e"
```
Comment (Sija) [inline src/uuid/json.cr]: Code-block ending marker (```) is missing here.

Comment (lachlan) [reply src/uuid/json.cr]: Wow, good pickup...fixed!

**Comment by RX14** [comment]
@lachlan you need to `make crystal` and use `bin/crystal tool format`, because your "run crystal tool format" commit was using the previous version's formatter and changed a bunch of unrelated formatting.

**Comment by lachlan** [comment]
@RX14 Thanks heaps! I didn't know what I was doing wrong, and did not even consider running `make crystal` again. Lesson learnt, build fixed.

**Comment by RX14** [comment]
Yeah, we should have a `make format` target which we can direct people to which is `.PHONY` and depends on `.build/crystal` I guess.

**Comment by wontruefree** [comment]
This looks like a good feature.  Thanks :)

**Pull request merged**